src/hotspot/share/opto/predicates.hpp

```diff
@@ -910,4 +918,70 @@ class Predicates : public StackObj {
 #endif // NOT PRODUCT
 };
 
+// This class checks whether a node is in the original loop body and not the cloned one.
+class NodeInOriginalLoopBody : public NodeInLoopBody {
+  const uint _first_node_index_in_cloned_loop_body;
+  const Node_List& _old_new;
+
+ public:
+  NodeInOriginalLoopBody(const uint first_node_index_in_cloned_loop_body, const Node_List& old_new)
+      : _first_node_index_in_cloned_loop_body(first_node_index_in_cloned_loop_body),
+        _old_new(old_new) {}
+  NONCOPYABLE(NodeInOriginalLoopBody);
+
+  // Check if 'node' is not a cloned node (i.e. "< _first_node_index_in_cloned_loop_body") and if we've created a
+  // clone from 'node' (i.e. _old_new entry is non-null). Then we know that 'node' belongs to the original loop body.
+  bool check(Node* node) const override {
+    if (node->_idx < _first_node_index_in_cloned_loop_body) {
+      Node* cloned_node = _old_new[node->_idx];
+      return cloned_node != nullptr && cloned_node->_idx >= _first_node_index_in_cloned_loop_body;
+    } else {
+      return false;
+    }
+  }
+};
+
+// Visitor to create Initialized Assertion Predicates at a target loop from Template Assertion Predicates from a source
+// loop. This visitor can be used in combination with a PredicateIterator.
+class AssertionPredicatesForLoop : public PredicateVisitor {
```

Comment:
I think this could have a more expressive name. It is a Visitor... hmm
Maybe `InitializedAssertionPredicatesFromTemplatesCreator`? Hmm not sure.

The current name suggests that it is just a collection of `AssertionPredicates`.

Reply:
As discussed offline, maybe `CreateAssertionPredicatesVisitor`?